File: BSharp/ClientApp/src/app/data/dto/local-user.ts
import { DtoForSaveKeyBase } from './dto-for-save-key-base';
import { RoleMembership, RoleMembershipForSave } from './role-membership';
import { SettingsForClient } from './settings';
import { DtoDescriptor } from './metadata';
import { TenantWorkspace } from '../workspace.service';
import { TranslateService } from '@ngx-translate/core';
import { DtoKeyBase } from './dto-key-base';

export class LocalUserForSave<TRoleMembership = RoleMembershipForSave> extends DtoForSaveKeyBase {
    Name: string;
    Name2: string;
    Email: string;
    Roles: TRoleMembership[];
    AgentId: number | string;
    Image: string;
}

export class LocalUser extends LocalUserForSave<RoleMembership> {
    ExternalId: string;
    ImageId: string;
    IsActive: boolean;
    LastAccess: string;
    CreatedAt: string;
    CreatedById: number | string;
    ModifiedAt: string;
    ModifiedById: number | string;
}

export class UserSettingsForClientForSave {
}

export class UserSettingsForClient {
    UserId: number;
    ImageId: string;
    Name: string;
    Name2: string;
}

const _select = ['', '2', '3'].map(pf => 'Name' + pf);
let _currentLang: string;
let _settings: SettingsForClient;
let _cache: DtoDescriptor;

export function metadata_LocalUser(ws: TenantWorkspace, trx: TranslateService, _subtype: string): DtoDescriptor {
  // Some global values affect the result, we check here if they have changed, otherwise we return the cached result
  if (trx.currentLang !== _currentLang && ws.settings !== _settings) {
    _currentLang = trx.currentLang;
    _settings = ws.settings;
    _cache = {
      select: _select,
      orderby: ws.isSecondaryLanguage ? `${_select[1]},${_select[0]}` : ws.isTernaryLanguage ? `${_select[2]},${_select[0]}` : _select[0],
      format: (item: DtoKeyBase) => ws.getMultilingualValueImmediate(item, _select[0]),
      properties: {
        Name: { control: 'text', label: trx.instant('Name') + ws.primaryPostfix },
        Name2: { control: 'text', label: trx.instant('Name') + ws.secondaryPostfix },
        Name3: { control: 'text', label: trx.instant('Name') + ws.ternaryPostfix },
        Email: { control: 'text', label: trx.instant('User_Email') },
        Agent: { control: 'navigation', label: trx.instant('User_Agent'), type: 'Custody', foreignKeyName: 'AgentId' },
        State: {
          control: 'state',
          label: trx.instant('State'),
          choices: ['New', 'Confirmed'],
          alignment: 'center',
          format: (c: string) => {
            switch (c) {
              case 'New': return trx.instant('User_New');
              case 'Confirmed': return trx.instant('User_Confirmed');
              default: return c;
            }
          },
          color: (c: string) => {
            switch (c) {
                case 'New': return '#6c757d';
                case 'Confirmed': return '#28a745';
                default: return c;
              }
          }
        },
        LastAccess: { control: 'datetime', label: trx.instant('User_LastActivity') },
        IsActive: { control: 'boolean', label: trx.instant('IsActive') },
        CreatedAt: { control: 'datetime', label: trx.instant('CreatedAt') },
        CreatedBy: { control: 'navigation', label: trx.instant('CreatedBy'), type: 'LocalUser', foreignKeyName: 'CreatedById' },
        ModifiedAt: { control: 'datetime', label: trx.instant('ModifiedAt') },
        ModifiedBy: { control: 'navigation', label: trx.instant('ModifiedBy'), type: 'LocalUser', foreignKeyName: 'ModifiedById' }
      }
    };
  }

  return _cache;
}

File: BSharp/Controllers/DTO/LocalUser.cs
using BSharp.Controllers.Misc;
using BSharp.Services.Utilities;
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using S = System.ComponentModel.DataAnnotations.Schema;

namespace BSharp.Controllers.DTO
{
    public class LocalUserForSave<TRoleMembership> : DtoForSaveKeyBase<int?>, IMultilingualName
    {
        [BasicField]
        [Required(ErrorMessage = nameof(RequiredAttribute))]
        [StringLength(255, ErrorMessage = nameof(StringLengthAttribute))]
        [MultilingualDisplay(Name = "Name", Language = Language.Primary)]
        public string Name { get; set; }

        [BasicField]
        [StringLength(255, ErrorMessage = nameof(StringLengthAttribute))]
        [MultilingualDisplay(Name = "Name", Language = Language.Secondary)]
        public string Name2 { get; set; }

        [Required(ErrorMessage = nameof(RequiredAttribute))]
        [EmailAddress(ErrorMessage = nameof(EmailAddressAttribute))]
        [StringLength(255, ErrorMessage = nameof(StringLengthAttribute))]
        [Display(Name = "User_Email")]
        public string Email { get; set; }

        [NavigationProperty(ForeignKey = nameof(RoleMembershipForQuery.UserId))]
        [Display(Name = "User_Roles")]
        public List<TRoleMembership> Roles { get; set; } = new List<TRoleMembership>();

        [ForeignKey]
        [Display(Name = "User_Agent")]
        public int? AgentId { get; set; }

        [S.NotMapped]
        [Display(Name = "Image")]
        public byte[] Image { get; set; }
    }

    public class LocalUserForSave : LocalUserForSave<RoleMembershipForSave> { }

    public class LocalUser<TRoleMembership> : LocalUserForSave<TRoleMembership>
    {
        public string ExternalId { get; set; }

        public string State { get; set; }

        public string ImageId { get; set; }

        [BasicField]
        [Display(Name = "IsActive")]
        public bool? IsActive { get; set; }

        [Display(Name = "User_LastActivity")]
        public DateTimeOffset? LastAccess { get; set; }

        [Display(Name = "CreatedAt")]
        public DateTimeOffset? CreatedAt { get; set; }

        [ForeignKey]
        [Display(Name = "CreatedBy")]
        public int? CreatedById { get; set; }

        [Display(Name = "ModifiedAt")]
        public DateTimeOffset? ModifiedAt { get; set; }

        [ForeignKey]
        [Display(Name = "ModifiedBy")]
        public int? ModifiedById { get; set; }
    }


    public class LocalUser : LocalUser<RoleMembership> { }

    public class LocalUserForQuery : LocalUser<RoleMembershipForQuery>, IAuditedDto
    {
        [NavigationProperty(ForeignKey = nameof(AgentId))]
        public AgentForQuery Agent { get; set; }

        [NavigationProperty(ForeignKey = nameof(CreatedById))]
        public LocalUserForQuery CreatedBy { get; set; }

        [NavigationProperty(ForeignKey = nameof(ModifiedById))]
        public LocalUserForQuery ModifiedBy { get; set; }
    }


    //////////////////// For Client

    /// <summary>
    /// Represents all user settings that a user can save: TODO
    /// </summary>
    public class UserSettingsForClientForSave
    {
    }

    /// <summary>
    /// Represents all user settings in a particular tenant
    /// </summary>
    public class UserSettingsForClient : UserSettingsForClientForSave
    {
        public int? UserId { get; set; }

        public string ImageId { get; set; }

        public string Name { get; set; }

        public string Name2 { get; set; }

        public Dictionary<string, string> CustomSettings { get; set; }
    }
}
